contracts/Chef.sol

```diff
@@ -0,0 +1,198 @@
+    ) external onlyRole(ALLOCATION_ROLE) nonDuplicated(newAddress) {
+        AllocationGroup memory group = AllocationGroup({
+            groupAddress: newAddress,
+            shares: newShares,
+            autodistribute: newAutoDistribute,
+            rewardDebt: 0,
+            pending: 0
+        });
+
+        _allocations.push(group);
+        _totalAllocPoint = _totalAllocPoint.add(group.shares);
+    }
+
+    function updateAllocationGroup(
+        uint256 agIndex,
+        uint256 shares,
+        bool newAutoDistribute
+    ) public onlyRole(ALLOCATION_ROLE) {
+        _totalAllocPoint = _totalAllocPoint.sub(_allocations[agIndex].shares).add(shares);
+        _allocations[agIndex].shares = shares;
+        _allocations[agIndex].autodistribute = newAutoDistribute;
+    }
+
+    function removeAllocationGroup(uint256 agIndex) external onlyRole(ALLOCATION_ROLE) {
+        require(agIndex < _allocations.length);
+        _totalAllocPoint = _totalAllocPoint.sub(_allocations[agIndex].shares);
+
+        _allocations[agIndex] = _allocations[_allocations.length - 1];
+        _allocations.pop();
+    }
+
+    function toggleRewards(bool isOn) external onlyRole(ALLOCATION_ROLE) {
+        _rewardsActive = isOn;
+        _lastRewardBlock = block.number;
+    }
+
+    //------------------------------------------------------Getters
+
+    function getAllocationGroups() public view returns (AllocationGroup[] memory) {
+        return _allocations;
+    }
+
+    function getTotalAllocationPoints() public view returns (uint256) {
+        return _totalAllocPoint;
+    }
+
+    //------------------------------------------------------Distribution
+
+    function viewPendingAllocations(uint256 agIndex) public view returns (uint256) {
+        AllocationGroup storage group = _allocations[agIndex];
+
+        return group.shares.mul(_accMETRICPerShare).div(ACC_METRIC_PRECISION).sub(group.rewardDebt);
+    }
+
+    function updateAccumulatedAllocations() public {
+        require(_rewardsActive, "Rewards are not active");
+        if (block.number <= _lastRewardBlock) {
+            return;
+        }
+
+        // TODO confirm budget is correct with assertions
+        // Not sure we can project emission rate over X years?
+        // Not entirely sure how to handle this, but we can at least try to make it work.
+        // ^^ will help with fuzz testing
+
+        uint256 blocks = block.number.sub(_lastRewardBlock);
+
+        uint256 accumulated = blocks.mul(METRIC_PER_BLOCK);
+
+        _accMETRICPerShare = _accMETRICPerShare.add(accumulated.mul(ACC_METRIC_PRECISION).div(_totalAllocPoint));
+        _lastRewardBlock = block.number;
+    }
+
+    function harvestAll() public {
+        for (uint256 i = 0; i < _allocations.length; i++) {
+            harvest(i);
+        }
+    }
+
+    function harvest(uint256 agIndex) public {
+        AllocationGroup storage group = _allocations[agIndex];
+
+        updateAccumulatedAllocations();
```

Comment:
Should this take `group` as a param?

Reply:
I mean we would still want to do a lookup for the share value, debt, etc within the contract because we can't trust the client to pass in their group and it's values accurately.

Reply:
yes -- see that it's tagged `public`, this is intended to be called from an external source (maybe another contract, maybe a user through a UI)